### PR TITLE
Introducing a LMMSE demosaicer

### PR DESCRIPTION
**Pull request by jenshannoschwalm**
[last edit 21/07/01]

Currently we clearly miss a demosaicer better suited for high ISO / noisy images.

There are a number of papers analysing image quality and S/N ratios, LMMSE is one of the algorithms with good results and a performance suited for realtime use as we require for dt.

LMMSE also handles raw moire patterns much better than PPG, RCD or AMaZE

The code for this dt adaption has been taken from rawtherapee/librtprocess, the algorithm is unchanged as i took that as the reference but has been tuned for performance quite a lot.

- no SSE specific code path. We can make sure for finite math in all sections of the code and thus can make use of aggressive compiler flags resulting in almost identical performance.
- the gamma correction is always used.
- internal tiling has been implemented for reduced memory footprint and thus improved performance as most data is kept in cpu cache.
- the original code used a lot of indirect pointers, this has been changed also for performance
- no OpenCL code yet. I am not sure if it's worth the effort as lmmse is likely not for general use ...
 

The performance is good now, for LMMSE default settings it is >2 times faster than AMaZE and just 70% of RCD cpu code.

### CODE REVIEW

File: src/iop/lmmse_demosaic.c [hunk 1]
```diff
@@ -0,0 +1,602 @@
...
+        rix[1] = qix[1] + rr * grp_width + ccc;
+        rix[4] = qix[4] + rr * grp_width + ccc;
+        rix[0][0] = 0.25f * (rix[4][ -2] + rix[4][ 2]) - 0.5f * (rix[4][ -1] + rix[4][0] + rix[4][ 1]);
+        rix[1][0] = 0.25f * (rix[4][-w2] + rix[4][w2]) - 0.5f * (rix[4][-w1] + rix[4][0] + rix[4][w1]);
+        rix[0][0] = limf(rix[0][0], -1.0f, 0.0f) + rix[4][0];
+        rix[1][0] = limf(rix[1][0], -1.0f, 0.0f) + rix[4][0];
+      }
+    }
+
+    // apply low pass filter on differential colors
```
Comment (Thanatomanic) [inline src/iop/lmmse_demosaic.c:317]: Be aware that this section makes heavy use of negative indices for arrays, which tools like `cppcheck` complain about. The code is valid though.

**Comment by dterrahe** [comment]
Duplications in gui_update can be avoided by having it call
`  gui_changed(self, NULL, NULL);
`
like is done in other similar modules. I think in the past gui_changed did less adjusting of widget visibility so there was less duplication, but now it would be worth simplifying.

**Comment by jenshannoschwalm** [comment]
@dterrahe fully agreed. The gui_update and gui_changed sections deserve de-duplication.

File: src/iop/lmmse_demosaic.c (hunk 1 above)
Comment (jenshannoschwalm) [reply src/iop/lmmse_demosaic.c:317]: ok. BTW after my fixing the currently still wrong refinement code (wrong input range) i would like to improve performance. From my understanding of commits in the rt codebase you seem to know the code pretty well. Interested in sort of collaboration on that? Via private mail or pixels.us ?

**Comment by aurelienpierre** [comment]
cool !

File: src/iop/lmmse_demosaic.c (hunk 1 above)
Comment (Thanatomanic) [reply src/iop/lmmse_demosaic.c:317]: Hi @jenshannoschwalm not nearly as well as Ingo, I've not dived into these kinds of algorithms before. But there could be a first time for everything. Let's use pixls for a chat.

**Comment by jenshannoschwalm** [comment]
Got it, the tiling did not work because of a very stupid bug ...

As the tiling is new code i would appreciate reviews by Ingo @heckflosse and @Thanatomanic ...

@aurelienpierre this could now be tested

**Comment by jenshannoschwalm** [comment]
A suggestion for the docs
************************************************************************
Darktable offers AMaZE and RCD as high quality demosaicers for low-nose / low-ISO images but tends to generate overshooting artefacts when applied to high-ISO / noisy images. For such images (or images with moire patterns) LMMSE is suited much better. LMMSE offers special refinement steps, use with care on raws with heavy chroma noise.
*************************************************************************

**Comment by jenshannoschwalm** [comment]
for the curious, hold on before testing. I am aware of some tiling-border problems that need fixing.

**Comment by jenshannoschwalm** [comment]
Fixed remaining tiling border issues and again tweaked performance by ~30%.

I think it is ok now for everyone to test. Have updated the introduction text too.

**Comment by johnny-bit** [comment]
Tested - can report that on high iso images the effect is i'd say pleasing and the speed's nice. Seems to result in less chroma blotches making even noisy images sorta easier to look at :) i need to check how it works with moire but it's nice!

**Comment by TurboGit** [comment]
Working, some points:
- you need to add the new file into po/POTFILES.in
- the result is finer than other algo, nice
- I found that it diminish the green more than other colors and so the tint derive to magenta a little (is that expected?).

**Comment by jenshannoschwalm** [comment]
> I found that it diminish the green more than other colors and so the tint derive to magenta a little (is that expected?).

I didn't recognize that. Any hint on what file / region i should look?

**Comment by TurboGit** [comment]
I took a very high iso image (12800). If you zoom at 200% you'll see that the green pixels are far more aggressively removed than using RCD or PPG for example. Then move out and look at the whole picture and changing from PPG to LMMSE you'll see that a more pinky tint is present in LMMSE.

**Comment by jenshannoschwalm** [comment]
>  If you zoom at 200% you'll see that the green pixels are far more aggressively removed

I checked many images and can not confirm so far. BUT i don't have many with such high iso and those i have are from other cameras. What i can confirm that there might be some "color shifting" but that is not always tending to red. I have images where there is heavy purple artefacts at edges that just go with lmmse.

Could you share such a problematic image?

This reminds me, maybe we would want a collection of "test images" ?

**Comment by jenshannoschwalm** [comment]
BTW i also checked all files i have vs RT output and could not see any visual difference.

**Comment by TurboGit** [comment]
Indeed, testing on some other pictures the color shit is not really visible. So it all depends on the RAW.

I cannot share the problematic images as not mine. Sorry.

**Comment by johnny-bit** [comment]
Just for clarification: is this LMMSE or DLMMSE? When looking for LMMSE papers my top result is DLMMSE paper.

If it's LMMSE... What's next on the menu? DLMMSE (which is supposed to be better in terms of output image quality)?

I also saw on your pixls.us thread also a paper about DDR/FR+ERP which looks nice (https://people.ee.ethz.ch/~timofter/#publications number 49 - with linked paper, src and demos) :)

**Comment by jenshannoschwalm** [comment]
Cited from the source :-)
```
    The lmmse code base used for the darktable port has been taken from rawtherapee derived librtprocess.
    Adaption for dt and tiling - hanno schwalm 06/2021

    LSMME demosaicing algorithm
    L. Zhang and X. Wu,
    Color demozaicing via directional Linear Minimum Mean Square-error Estimation,
    IEEE Trans. on Image Processing, vol. 14, pp. 2167-2178, Dec. 2005.

    Adapted to RawTherapee by Jacques Desmis 3/2013
    Improved speed and reduced memory consumption by Ingo Weyrich 2/2015

```

i just kept is as LMMSE as that also has been used in RT

**Comment by johnny-bit** [comment]
Ah, so it IS DLMMSE!

**Review by TurboGit — APPROVED** [review]
That's good to me. Thanks !

**Comment by elstoc** [comment]
@jenshannoschwalm: documentation [here](https://github.com/darktable-org/dtdocs/commit/287a06d3805fb3dac6368e13811a0b81d4aa85f3). Please can you give it a quick review. Could we add any more detail to the lmmse refine option (recommendations for use etc)?

**Pull request merged**